source/predicates/array.ts

```diff
@@ -156,6 +156,6 @@ export class ArrayPredicate<T = unknown> extends Predicate<T[]> {
 					return false;
 				}
 			}
-		});
+		}) as unknown as ArrayPredicate<U>;
```

Comment:
Why this? Is there really no way to have the TS handle the type without forcing it?

Reply:
Not without higher-kinded types, I don't think. The return type of `addValidator` is `this`. Which means `addValidator` of `ArrayPredicate<T>` is `ArrayPredicate<T>`. We could probably make `addValidator` take a generic parameter `U` and have it return `Predicate<T & U>`, but then we'd be downcasting from `ArrayPredicate` to `Predicate`. I can push a change to make this `ArrayPredicate<any>` though so there isn't the double as-cast.

Higher-kinded types might conceivably land in a distant typescript future: https://github.com/microsoft/TypeScript/pull/40368

I've added a todo which will trigger on the next major version of typescript, in case they do make it in. There are some workarounds in libraries like fp-ts, but I don't think the overhead/complexity is worth it in this case since it's internal to this library.